### DISCUSSION TITLE
help with dockerized nextjs app with nginx/apache

### DISCUSSION THREAD

**Discussion by Lueton** [vercel/next.js#51577] [Help]
### Summary

We recently upgraded our old nextjs to the latest version. Turns out nothing is working anymore. We tried a lot of thing but were not able to make it work. So we tried to create a fresh nextjs app (`npx create-next-app@latest`) and test it. The only thing we modified was to set `output: 'standalone'` in our `next.config.js`. To deploy the app we use the `with-docker` example provided by nextjs without any modification. But it still does not work neither with nginx nor with apache. Its seems like the app ist just not responding. If we test it locally with docker desktop it is working. We are thankful for any help. Please let us know if you need additional informations.

`os: ubuntu 22.04`
`nginx: 1.18.0`
`podman: 3.4.4`
`next: 13.4.6`
`react: 18.2.0`
`react-dom: 18.2.0`

Dockerfile
```
FROM node:18-alpine AS base

# Install dependencies only when needed
FROM base AS deps
# Check https://github.com/nodejs/docker-node/tree/b4117f9333da4138b03a546ec926ef50a31506c3#nodealpine to understand why libc6-compat might be needed.
RUN apk add --no-cache libc6-compat
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json yarn.lock* package-lock.json* pnpm-lock.yaml* ./
RUN \
  if [ -f yarn.lock ]; then yarn --frozen-lockfile; \
  elif [ -f package-lock.json ]; then npm ci; \
  elif [ -f pnpm-lock.yaml ]; then yarn global add pnpm && pnpm i --frozen-lockfile; \
  else echo "Lockfile not found." && exit 1; \
  fi


# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
COPY --from=deps /app/node_modules ./node_modules
COPY . .

# Next.js collects completely anonymous telemetry data about general usage.
# Learn more here: https://nextjs.org/telemetry
# Uncomment the following line in case you want to disable telemetry during the build.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN yarn build

# If using npm comment out above and use below instead.
# RUN npm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app

ENV NODE_ENV production
# Uncomment the following line in case you want to disable telemetry during runtime.
# ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Automatically leverage output traces to reduce image size
# https://nextjs.org/docs/advanced-features/output-file-tracing
COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs

EXPOSE 3000

ENV PORT 3000

CMD ["node", "server.js"]
```
Nginx (try 1)
```
server {
    listen 80;
    server_name localhost:3000;

    location / {
        proxy_pass http://localhost:3000;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
        proxy_cache_bypass $http_upgrade;
    }
}
```
Nginx (try 2)
```
upstream nextjs_upstream {
  server nextjs:3000;
}

server {
  listen 80 default_server;

  server_name _;

  server_tokens off;

  proxy_http_version 1.1;
  proxy_set_header Upgrade $http_upgrade;
  proxy_set_header Connection 'upgrade';
  proxy_set_header Host $host;
  proxy_cache_bypass $http_upgrade;

  location / {
    proxy_pass http://nextjs_upstream;
  }
}
```
`error 1: connect() failed (111: Unknown error) while connecting to upstream` 
`error 2: no live upstreams while connecting to upstream`

**Comment by icyJoseph** [comment]
Could it be something with ipv6? or maybe 0.0.0.0 instead of localhost?

**Reply by Lueton** [reply]
checked your idea with `0.0.0.0` and disabled ipv6. Did not help so far.

**Comment by gerardnico** [comment]
Delete the proxy, no nginx, no apache.

**Reply by Lueton** [reply]
I like the idea! Disabled nginx with `systemctl stop nginx`. Should be enough right? Anyway there is still no answer from the container, even if i try it with `localhost:3000` directly on our server (`Failed to connect to localhost:3000: Connection timed out`). Podman logs for container just returns `Listening on port 3000 url: http://2aa4843152cb:3000`. No further errors. Port mapping is `0.0.0.0:3000->3000/tcp`

**Comment by Lueton** [comment]
Well, I already had a feeling that there must be something obvious we missed.  Turns out the we need to run the container with the `--net host` flag, which was somehow not required on our old infra. This leads me to the conclusion that this has nothing to do with nextjs.
